server/src/test/java/com/angelomelonas/grpcwebchat/integration/ChatServiceTest.java

```diff
@@ -0,0 +1,260 @@
+import java.util.UUID;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class ChatServiceTest {
+
+    private ChatServiceTestClient chatServiceTestClient;
+
+    @Before
+    public void resetChatClient() {
+        chatServiceTestClient = new ChatServiceTestClient();
+    }
+
+    @After
+    public void cleanUp() throws InterruptedException {
+        chatServiceTestClient.shutdown();
+    }
+
+    @Test
+    public void authenticateTest() {
+        AuthenticationResponse authenticationResponse = chatServiceTestClient.authenticate(AuthenticationRequest.newBuilder().build());
+
+        Assert.assertNotNull(authenticationResponse.getUuid());
+    }
+
+    @Test
+    public void subscribeTest() {
+        AuthenticationResponse authenticationResponse = chatServiceTestClient.authenticate(AuthenticationRequest.newBuilder().build());
+
+        UUID uuid = UUID.fromString(authenticationResponse.getUuid());
+        String username = "TestUsername" + randomSuffixGenerator();
+
+        SubscriptionRequest subscriptionRequest = SubscriptionRequest.newBuilder()
+                .setUuid(String.valueOf(uuid))
+                .setUsername(username)
+                .build();
+
+        StreamObserverTestHelper<Message> streamObserver = new StreamObserverTestHelper();
+
+        chatServiceTestClient.subscribe(subscriptionRequest, streamObserver);
+        // Wait for the server to acknowledge the subscription.
+        Message messageResponse = streamObserver.waitForOnNext().get();
+
+        Assert.assertEquals("Server", messageResponse.getUsername());
+        Assert.assertEquals("Client subscribed successfully.", messageResponse.getMessage());
+
+        UnsubscriptionRequest unsubscriptionRequest = UnsubscriptionRequest.newBuilder()
+                .setUuid(String.valueOf(uuid))
+                .build();
+
+        chatServiceTestClient.unsubscribe(unsubscriptionRequest);
+    }
+
+    @Test
+    public void unsubscribeTest() {
+        AuthenticationResponse authenticationResponse = chatServiceTestClient.authenticate(AuthenticationRequest.newBuilder().build());
+
+        UUID uuid = UUID.fromString(authenticationResponse.getUuid());
+        String username = "TestUsername" + randomSuffixGenerator();
+
+        SubscriptionRequest subscriptionRequest = SubscriptionRequest.newBuilder()
+                .setUuid(String.valueOf(uuid))
+                .setUsername(username)
+                .build();
+
+        StreamObserverTestHelper<Message> streamObserver = new StreamObserverTestHelper();
+
+        chatServiceTestClient.subscribe(subscriptionRequest, streamObserver);
+        // Wait for the server to acknowledge the subscription.
+        streamObserver.waitForOnNext();
+
+        UnsubscriptionRequest unsubscriptionRequest = UnsubscriptionRequest.newBuilder()
+                .setUuid(String.valueOf(uuid))
+                .build();
+
+        Message unsubscribeMessage = chatServiceTestClient.unsubscribe(unsubscriptionRequest);
```

Comment:
Maybe consider using a different response message for subscribe and unsubscribe, reusing "Message" here just doesn't look good.

You can have a generic Answer message for subscribe and unsubscribe

Reply:
the client does not need to know the username ... all the client wants to know whether or not the (un)subscribe was successful or not

Reply:
I will make these changes in an upcoming PR.